eta/core/serial.py

```diff
@@ -670,32 +705,325 @@ def from_dict(cls, d):
+    def __delitem__(self, idx):
+        etau.delete_file(self._ele_path(idx))
+        super(DirectoryContainer, self).__delitem__(idx)
+
+    def add(self, ele):
+        self.append(ele)
+
+    def add_container(self, container):
+        '''Add a another containers elements to this one. Takes any valid
+        Container subclass (Not just DirectoryContainer subclasses).
+
+        Args:
+            container (Container): the container to add
+        '''
+        for ele in container:
+            self.add(ele)
+
+    def copy(self, new_ele_dir):
+        '''Copy this container deeply.
+
+        Args:
+            new_ele_dir (str): path the store the copy's elements
+
+        Returns:
+            DirectoryContainer
+        '''
+        try:
+            etau.delete_dir(new_ele_dir)
+        except OSError:
+            pass
+        container = self.__class__(**{self._ELE_DIR_ATTR: new_ele_dir})
+        container.add_container(self)
+        return container
+
+    def append(self, ele):
+        '''Append an element.'''
+        self[len(self)] = ele
+
+    def filter_elements(self, filters, match=any):
+        '''Removes elements that don't match the given filters from the
+        container.
+
+        Args:
+            filters: a list of functions that accept elements and return
+                True/False
+            match: a function (usually `any` or `all`) that accepts an iterable
+                and returns True/False. Used to aggregate the outputs of each
+                filter to decide whether a match has occurred. The default is
+                `any`
+        '''
+        self._filter_elements(filters, match)
+
+    def delete_inds(self, inds):
+        '''Deletes the elements from the container with the given indices.
+
+        Args:
+            inds: a list of indices of the elements to delete
+        '''
+        for idx in sorted(inds, reverse=True):
+            del self[idx]
+
+    def keep_inds(self, inds):
+        '''Keeps only the elements in the container with the given indices.
+
+        Args:
+            inds: an iterable of indices of the elements to keep
+        '''
+        inds = set(inds)
+        for idx, path in self._files():
+            if idx not in inds:
+                del self[idx]
+
+    def extract_inds(self, new_ele_dir, inds):
+        '''Creates a new container having only the elements with the given
+        indices.
+
+        Args:
+            inds: a list of indices of the elements to keep
+
+        Returns:
+            DirectoryContainer
+        '''
+        inds = set(inds)
+        container = copy.deepcopy(self)
+        setattr(container, self._ELE_DIR_ATTR, new_ele_dir)
+        for idx, ele in self:
+            if idx in inds:
+                container.add(ele)
+        return container
+
+    def clear(self):
+        '''Deletes all elements from the container.'''
+        super(DirectoryContainer, self).clear()
+        etau.delete_dir(self.__elements_dir__)
```

Comment:
Hmm, what if you have other stuff in that directory?

Reply:
Will do.

Reply:
Yes, I now understand this better and agree.  The `backing_dir` should be solely managed by the `BigContainer` class, and should start out empty.  (This would be similar to `LabeledDataset`.) The general workflow would be that you create an empty `BigContainer` and then copy over initial files from somewhere else, e.g. a `LabeledDataset`. I think that you should add the following methods for a `BigContainer`:
- create a `BigContainer` from a list of existing paths to files.  This would ideally copy the files using disk operations to the `backing_dir`, rather then reading them into memory and writing them into the `backing_dir`.
- create a `BigVideoSetLabels` from a `LabeledVideoDataset`, which similarly copies over all of the labels files from the `LabeledVideoDataset`.  (Similarly for a `BigImageSetLabels`.)

Reply:
I would do this as `BigContainer.from_files()` and `VideoSetLabels.from_labeled_video_dataset()` and `BigVideoSetLabels.from_labeled_video_dataset()`

Same for `ImageSetLabels` and `LabeledImageDataset`

Reply:
Updates made!

`from_paths`, `from_dir`, and `from_zip` are now present. I've tested some, but not enough.